app/scripts/modules/core/navigation/urlBuilder.service.ts

```diff
@@ -0,0 +1,443 @@
+import {isDate, isObject, isUndefined} from 'lodash';
+import {module} from 'angular';
+
+import {IStateService} from 'angular-ui-router';
+
+export interface IUrlBuilderInput {
+  account: string;
+  application: string;
+  cluster: string;
+  detail: string;
+  instanceId: string;
+  loadBalancer: string;
+  name: string;
+  project: string;
+  provider: string;
+  region: string;
+  serverGroup: string;
+  stack: string;
+  taskId: string;
+  type: string;
+  vpcId: string;
+}
```

Comment:
Aside from `type` and `name` (and possibly including `type` and `name`), all of these fields are probably optional.